Add initial argocd helm deployment docs (#55835)
gravitational/teleport: docs/pages/admin-guides/deploy-a-cluster/helm-deployments/argocd-helm.mdx

---
title: Guides for running Teleport using Helm via ArgoCD
description: How to install and configure Teleport in Kubernetes using Helm and ArgoCD
---

Teleport can provide secure, unified access to your Kubernetes clusters. This
guide will show you how to deploy Teleport Kubernetes agent on a Kubernetes cluster using Helm
and ArgoCD.

## How it works

> AArgo CD is a declarative, GitOps continuous delivery tool for Kubernetes. This is used to orchestrate large deployments, and avoid the Kubernetes resources to drift from the desired deployment.

Reviewer: ```suggestion
Argo CD is a declarative, GitOps continuous delivery tool for Kubernetes. This is used to orchestrate large deployments, and avoid the Kubernetes resources to drift from the desired deployment.
```

> Teleport has [an official Helm chart (`teleport-kube-agent`)](../../../reference/helm-reference/teleport-kube-agent.mdx) that deploys a Teleport agent in a Kubernetes cluster. The agent can be configured to run several services, but by default it runs the `kubernetes_sevrice` to provide access to the Kubernetes API via Teleport.

Reviewer: ```suggestion
Teleport has [an official Helm chart (`teleport-kube-agent`)](../../../reference/helm-reference/teleport-kube-agent.mdx) that deploys a Teleport agent in a Kubernetes cluster. The agent can be configured to run several services, but by default it runs the `kubernetes_service` to provide access to the Kubernetes API via Teleport.
```

This guide leverages ArgoCD's native Helm support to deploy the Teleport agent using the `teleport-kube-agent` Helm chart.

## Prerequisites

- A registered domain name. This is required for Teleport to set up TLS via

>   Let's Encrypt and for Teleport clients to verify the Proxy Service host.

Reviewer: What is this for? We're not deploying the control plane?

- An existing Kubernetes cluster you wish to provide access to via Teleport.

- At least one Teleport Agent running in your cluster publicly reachable with
  valid TLS certificates

 - (!docs/pages/includes/tctl.mdx!)

- An existing ArgoCD instance (version 2.10 or greater) that can deploy to the
  above Kubernetes cluster.

- The `tsh` client tool v(=teleport.version=)+ installed on your workstation.
  You can download this from our [installation page](../../../installation/installation.mdx).
  
## Step 1/2. Generate a join token

Teleport agents use a join token to obtain certificates and connect to Teleport. See [joining docs](../../../reference/join-methods.mdx) for more information.
The token must be valid only on first join, the Teleport Kube agent will store its
certificates in Kubernetes and won't need a token to join again in the future.
In this section, we will create a token for the agent to join the Teleport cluster.
```code
$ tctl tokens add --type=kube,app --ttl=5m
```

You can specify the following token types:

(!docs/pages/includes/token-types.mdx!)

Ensure you update the values file you use for the helm chart

> to include the required roles.

Reviewer: What does this mean? As a user/reader, it's not clear. A specific example would help.

See the `teleport-kube-agent` [chart
reference](../../../reference/helm-reference/teleport-kube-agent.mdx#roles) for the
roles and token types that the chart supports.

## Step 2/2. Configure and deploy the `teleport-kube-agent` Helm chart via ArgoCD

1. Create a namespace for Teleport and configure its Pod Security Admission,
   which enforces security standards on pods in the namespace:

   ```code
   $ kubectl create namespace teleport
   namespace/teleport created
   
   $ kubectl label namespace teleport 'pod-security.kubernetes.io/enforce=baseline'
   namespace/teleport labeled
   ```

2. Create a new ArgoCD application using the following as a template.

``` yaml
project: default
source:
  repoURL: 'https://charts.releases.teleport.dev'
  targetRevision: (=teleport.version=)
  helm:
    values: |-

      roles: kube,app,discovery
      authToken: $YOUR_AUTH_TOKEN
      proxyAddr: $YOUR_PROXY_ADDRESS
      kubeClusterName: $YOUR_KUBE_CLUSTER_NAME

      highAvailability:
          replicaCount: 2
          podDisruptionBudget:
              enabled: true
              minAvailable: 1
  chart: teleport-kube-agent
destination:
  server: 'https://kubernetes.default.svc'

>   namespace: $YOUR_TARGET_NAMESPACE

Reviewer: What's the namespace here? We're asking people to create "teleport" namespace above, should this just be "teleport"?

# This section is used to allow the teleport-kube-agent-updater to update the agent
# without ArgoCD reverting the update.
ignoreDifferences:
  - group: apps
    kind: StatefulSet
    name: $YOUR_APPLICATION_NAME

>     namespace: $YOUR_TARGET_NAMESPACE

Reviewer: Same comment about namespace as above.

  
- ArgoCD version 2.10 or greater.

## Step 1/2. Install Teleport

1. Install the `teleport-cluster` Helm chart, which deploys the Teleport Auth
   Service and Proxy Service on your Kubernetes cluster.

### Install the `teleport-cluster` Helm chart

1. Create a namespace for Teleport and configure its Pod Security Admission,
   which enforces security standards on pods in the namespace:

   ```code
   $ kubectl create namespace teleport-cluster
   namespace/teleport-cluster created
   
   $ kubectl label namespace teleport-cluster 'pod-security.kubernetes.io/enforce=baseline'
   namespace/teleport-cluster labeled
   ```

2. Navigate to the 'Application Page in ArgoCD', then select 'New App'.

![ArgoCD application page](../../../../img/helm/argocd/application-page.png)

3. Then select 'Edit as YAML'

![ArgoCD application page](../../../../img/helm/argocd/edit-yaml.png)

Example teleport-cluster configuration
``` yaml
apiVersion: argoproj.io/v1alpha1
kind: Application
metadata:
  name: teleport-cluster
spec:
  destination:
    name: ''
    namespace: teleport
    server: 'https://kubernetes.default.svc'
  source:
    path: ''
    repoURL: 'https://charts.releases.teleport.dev'

>     targetRevision: 17.4.10

Reviewer 2: If you need to specify versions in the docs, always use (=teleport.version=), else the version will stay stale. For example here, we are telling users to install a highly vulnerable teleport version.

as a member of the `viewers` group, this group does not yet have permissions
within its Kubernetes cluster. To assign these permissions, create a Kubernetes
`RoleBinding` or `ClusterRoleBindings` that grants permission to the `viewers`
group.

1. Create a file called `viewers-bind.yaml` with the following contents:

   ```yaml
   apiVersion: rbac.authorization.k8s.io/v1
   kind: ClusterRoleBinding
   metadata:
     name: viewers-crb
   subjects:
   - kind: Group
     # Bind the group "viewers" to the kubernetes_groups assigned in the "kube-access" role
     name: viewers
     apiGroup: rbac.authorization.k8s.io
   roleRef:
     kind: ClusterRole
     # "view" is a default ClusterRole that grants read-only access to resources
     # See: https://kubernetes.io/docs/reference/access-authn-authz/rbac/#user-facing-roles
     name: view
     apiGroup: rbac.authorization.k8s.io
   ```

1. Apply the `ClusterRoleBinding` with `kubectl`:

   ```code
   $ kubectl apply -f viewers-bind.yaml
   ```

Now any Teleport user you grant the `kube-access` role to will be able to see the
Kubernetes cluster and view its content as a member of the `viewers` group.

## Next steps

>  teleport-kube-agent Helm reference, the Teleport kube access RBAC docs, and maybe the argoCD ones.

Reviewer: What's this line about?

   ```code
   $ kubectl create namespace teleport
   namespace/teleport created
   
   $ kubectl label namespace teleport 'pod-security.kubernetes.io/enforce=baseline'
   namespace/teleport labeled
   ```

2. Create a new ArgoCD application using the following as a template.

``` yaml
project: default
source:
  repoURL: 'https://charts.releases.teleport.dev'
  targetRevision: (=teleport.version=)
  helm:
    values: |-
      roles: kube,app
      authToken: $YOUR_AUTH_TOKEN
      proxyAddr: $YOUR_PROXY_ADDRESS
      kubeClusterName: $YOUR_KUBE_CLUSTER_NAME

      highAvailability:
          replicaCount: 2
          podDisruptionBudget:
              enabled: true
              minAvailable: 1
  chart: teleport-kube-agent
destination:
  server: 'https://kubernetes.default.svc'
  namespace: teleport
# This section is used to allow the teleport-kube-agent-updater to update the agent
# without ArgoCD reverting the update.
ignoreDifferences:
  - group: apps
    kind: StatefulSet
    name: $YOUR_APPLICATION_NAME
    namespace: teleport
    jqPathExpressions:
      - '.spec.template.spec.containers[] | select(.name == "teleport").image'
```

> 3. Sync your changes to apply the configuration using `argocd app sync $YOUR_APPLICATION_NAME`

Reviewer: Can you format the command properly?
```suggestion
3. Sync your changes to apply the configuration using the following command:
```bash
$ argocd app sync $YOUR_APPLICATION_NAME
```
```

4. To verify setup, navigate to the 'Resources' page in your Teleport

>    cluster to confirm the Kubernetes cluster is registered.

Reviewer: I think a step is missing after this, how do I now give access to this cluster to my users? I don't mean "step 5" here, I mean top level "Step 3" of the guide. I've registered my agent, now how do I access it.

Can you add a section showing the role configuration and connecting to this cluster via Teleport. Without this the guide feels incomplete.

```bash
$ argocd app sync $YOUR_APPLICATION_NAME
```
4. To verify setup check the readiness of the pods in the teleport namespace.

> ```bash
> $ kubectl --namespace teleport get all
> ```

Reviewer 2: Our docs engine requires us to use the `code` block for copy-paste to work properly. Commands should reduce the output to the strict minimum (all -> pods) and if you are expecting the user to validate that something is working properly, it's better to show them an example of what they should see (the command output).

````suggestion
```code
$kubectl -n teleport get pods

NAME                                           READY   STATUS    RESTARTS   AGE
<your-agent-release>-0                         1/1     Running   0          14m
<your-agent-release>-1                         1/1     Running   0          14m
<your-agent-release>-updater-d9f97f5dd-v57g9   1/1     Running   0          16m
```
````
